### PR TITLE
Check status of connection after PQconnectStartParams

### PR DESCRIPTION
**Pull request by knizhnik**
## Problem

See https://github.com/neondatabase/cloud/issues/14289

## Summary of changes

Check connection status after calling PQconnectStartParams

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by kelvich** [comment]
Thank you, good catch!

Looking at `PQstatus` code it seem that previous check is not needed and can be omitted (`PQerrorMessage` handles null pointer too).

```
ConnStatusType
PQstatus(const PGconn *conn)
{
	if (!conn)
		return CONNECTION_BAD;
	return conn->status;
}
```

Also that is a good example of how not to write commit messages: you are describing code change as such and that is visible from the PR itself in this case. What is not obvious is why this change is done (to fix `cannot wait on socket event without a socket` error)

**Comment by hlinnaka** [comment]
A test would be nice too

**Comment by knizhnik** [comment]
> Also that is a good example of how not to write commit messages: you are describing code change as such and that is visible from the PR itself in this case. What is not obvious is why this change is done (to fix `cannot wait on socket event without a socket` error)

Sorry, writing good commit messages and comments was alway my problem.
But generally I thought that PR description should explain the source problem and how to reproduce it while commit message should explain how it is fixed. Should commit message be self-explained and contain both: description of the problem and its fix? Not obvious to me...

File: pgxn/neon/libpagestore.c [hunk 1]
```diff
@@ -432,7 +432,11 @@ pageserver_connect(shardno_t shard_no, int elevel)
 			neon_shard_log(shard_no, elevel, "Failed to connect to pageserver: out of memory");
 			return false;
 		}
-
+		if (PQstatus(shard->conn) == CONNECTION_BAD)
+		{
+			neon_shard_log(shard_no, elevel, "Failed to connect to pageserver: %s", PQerrorMessage(shard->conn));
+			return false;
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon/libpagestore.c]: Thank you.
Done.
By there way, I wonder why you have named this function `CLEANUP_AND_DISCONNECT` (all uppercases).
It seems to violate any Postgres style guide.
Why this function is so critical that it it necessary "caps-lock" it?

**Comment by knizhnik** [comment]
> A test would be nice too

Yeh, it is non trivial. The errors seems to handle after OOM.
I do not know some convenient and reliable way to cause permanent OOM.
Certainly it is possible to execute queries creating huge strings and so consume all memory.
But once this query is canceled all this memory will be deallocated.

Also it OOM should happen in proper place. If `PQconnectStartParams` failed to allocate PGconn because of lack of memory, then it will return NULL which is already handled.

**Comment by hlinnaka** [comment]
> > A test would be nice too
> 
> Yeh, it is non trivial. The errors seems to handle after OOM.

The libpq docs say:

> To begin a nonblocking connection request, call PQconnectStart or [PQconnectStartParams](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PQCONNECTSTARTPARAMS). If the result is null, then libpq has been unable to allocate a new PGconn structure. Otherwise, a valid PGconn pointer is returned (though not yet representing a valid connection to the database). Next call PQstatus(conn). If the result is CONNECTION_BAD, the connection attempt has already failed, typically because of invalid connection parameters.

Note the last sentence. So OOM is one way to hit this, but invalid connection parameters is much simpler to do in a test.

**Comment by knizhnik** [comment]
> > > A test would be nice too
> > 
> > 
> > Yeh, it is non trivial. The errors seems to handle after OOM.
> 
> The libpq docs say:
> 
> > To begin a nonblocking connection request, call PQconnectStart or [PQconnectStartParams](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PQCONNECTSTARTPARAMS). If the result is null, then libpq has been unable to allocate a new PGconn structure. Otherwise, a valid PGconn pointer is returned (though not yet representing a valid connection to the database). Next call PQstatus(conn). If the result is CONNECTION_BAD, the connection attempt has already failed, typically because of invalid connection parameters.
> 
> Note the last sentence. So OOM is one way to hit this, but invalid connection parameters is much simpler to do in a test.

Sorry, I too harried to merge this PR.
Test is added I separate PR:
https://github.com/neondatabase/neon/pull/8231

**Pull request merged**